File: scratch/test_calc.py
import json
import math

def calculate_tax(tax_base):
    # 종합소득세 세율 구간 (2024~2025 귀속)
    brackets = [
        (14000000, 0.06, 0),
        (50000000, 0.15, 1260000),
        (88000000, 0.24, 5760000),
        (150000000, 0.35, 15440000),
        (300000000, 0.38, 19940000),
        (500000000, 0.40, 25940000),
        (1000000000, 0.42, 35940000),
        (float('inf'), 0.45, 65940000)
    ]
    if tax_base <= 0:
        return 0
    for limit, rate, deduction in brackets:
        if tax_base <= limit:
            return (tax_base * rate) - deduction
    return 0

def truncate_won(amount):
    return math.floor(amount / 10) * 10

def calculate_tax_simulation(accumulated_income, input_months, allowance_count, occupation_code):
    months = max(1, min(12, input_months))
    income = max(0.0, accumulated_income)
    dependents = max(0, allowance_count)
    
    # 2025년 귀속 경비율.json 로드
    with open('지식_변환/JSON/2025년 귀속 경비율.json', 'r', encoding='utf-8') as f:
        data = json.load(f)
    
    rows = data.get('데이터', [])
    simple_base_rate = 0.0
    simple_excess_rate = 0.0
    standard_rate = 0.0
    occupation_name = '미등록 업종'
    
    for row in rows[2:]:
        if row[0].strip() == occupation_code:
            occupation_name = row[1].strip()
            simple_base_rate = float(row[2].strip()) if row[2] else 0.0
            simple_excess_rate = float(row[3].strip()) if len(row) > 3 and row[3] else 0.0
            if simple_excess_rate == 0.0:
                simple_excess_rate = simple_base_rate
            standard_rate = float(row[4].strip()) if len(row) > 4 and row[4] else 0.0
            break
            
    simple_base_rate_pct = simple_base_rate / 100.0
    simple_excess_rate_pct = simple_excess_rate / 100.0
    
    # 미래 예측 (연환산 연소득)
    annual_estimated_income = (income / months) * 12
    
    # 필요경비 (단순경비율)
    if annual_estimated_income <= 40000000:
        estimated_expense = annual_estimated_income * simple_base_rate_pct
    else:
        estimated_expense = (40000000 * simple_base_rate_pct) + ((annual_estimated_income - 40000000) * simple_excess_rate_pct)
        
    estimated_business_income = annual_estimated_income - estimated_expense
    
    # 인적공제
    deduction = (dependents + 1) * 1500000.0
    
    # 과세표준
    tax_base = estimated_business_income - deduction
    if tax_base < 0:
        tax_base = 0
        
    # 산출세액
    estimated_calculated_tax = calculate_tax(tax_base)
    estimated_local_tax = estimated_calculated_tax * 0.1
    
    final_annual_income_tax = truncate_won(estimated_calculated_tax)
    final_annual_local_tax = truncate_won(estimated_local_tax)
    final_annual_total_tax = final_annual_income_tax + final_annual_local_tax
    
    # 기납부세액
    paid_income_tax = truncate_won(income * 0.03)
    paid_local_tax = truncate_won(income * 0.003)
    paid_total_withholding = paid_income_tax + paid_local_tax
    
    annual_estimated_withholding_income = truncate_won(annual_estimated_income * 0.03)
    annual_estimated_withholding_local = truncate_won(annual_estimated_income * 0.003)
    annual_estimated_total_withholding = annual_estimated_withholding_income + annual_estimated_withholding_local
    
    # 예상 결과
    expected_refund_or_payment = annual_estimated_total_withholding - final_annual_total_tax
    
    # 저축 넛지
    monthly_reserve = 0.0
    if expected_refund_or_payment < 0:
        additional_payment = abs(expected_refund_or_payment)
        remaining_months = 12 - months
        if remaining_months > 0:
            monthly_reserve = truncate_won(additional_payment / remaining_months)
        else:
            monthly_reserve = truncate_won(additional_payment)
            
    return {
        'annual_estimated_income': annual_estimated_income,
        'estimated_expense': estimated_expense,
        'estimated_business_income': estimated_business_income,
        'tax_base': tax_base,
        'annual_income_tax': final_annual_income_tax,
        'annual_local_tax': final_annual_local_tax,
        'annual_total_tax': final_annual_total_tax,
        'paid_total_withholding': paid_total_withholding,
        'annual_estimated_total_withholding': annual_estimated_total_withholding,
        'expected_refund_or_payment': expected_refund_or_payment,
        'monthly_reserve': monthly_reserve,
        'occupation_name': occupation_name,
        'simple_base_rate': simple_base_rate,
        'simple_excess_rate': simple_excess_rate,
        'standard_rate': standard_rate
    }

# 시나리오 1 테스트
res1 = calculate_tax_simulation(30000000, 6, 0, '940909')
print("[시나리오 1 - 파이썬 검증]")
print(f"업종명: {res1['occupation_name']}")
print(f"연환산 수입: {res1['annual_estimated_income']}")
print(f"추정 필요경비: {res1['estimated_expense']}")
print(f"추정 사업소득금액: {res1['estimated_business_income']}")
print(f"추정 과세표준: {res1['tax_base']}")
print(f"추정 종합소득세: {res1['annual_income_tax']}")
print(f"추정 지방소득세: {res1['annual_local_tax']}")
print(f"연 추정 기납부 3.3% 세액: {res1['annual_estimated_total_withholding']}")
print(f"최종 예상 환급/납부액: {res1['expected_refund_or_payment']}")
print(f"월 권장 저축액: {res1['monthly_reserve']}")

assert res1['annual_estimated_income'] == 60000000.0
assert res1['estimated_expense'] == 35580000.0
assert res1['estimated_business_income'] == 24420000.0
assert res1['tax_base'] == 22920000.0
assert res1['annual_income_tax'] == 2178000.0
assert res1['annual_local_tax'] == 217800.0
assert res1['expected_refund_or_payment'] == -415800.0
assert res1['monthly_reserve'] == 69300.0
print("시나리오 1 파이썬 검증 통과!")

# 시나리오 2 테스트
res2 = calculate_tax_simulation(10000000, 3, 1, '940918')
print("\n[시나리오 2 - 파이썬 검증]")
print(f"업종명: {res2['occupation_name']}")
print(f"연환산 수입: {res2['annual_estimated_income']}")
print(f"추정 필요경비: {res2['estimated_expense']}")
print(f"추정 사업소득금액: {res2['estimated_business_income']}")
print(f"추정 과세표준: {res2['tax_base']}")
print(f"추정 종합소득세: {res2['annual_income_tax']}")
print(f"추정 지방소득세: {res2['annual_local_tax']}")
print(f"최종 예상 환급/납부액: {res2['expected_refund_or_payment']}")

assert res2['annual_estimated_income'] == 40000000.0
assert res2['estimated_expense'] == 31760000.0
assert res2['estimated_business_income'] == 8240000.0
assert res2['tax_base'] == 5240000.0
assert res2['annual_income_tax'] == 314400.0
assert res2['annual_local_tax'] == 31440.0
assert res2['expected_refund_or_payment'] == 974160.0
print("시나리오 2 파이썬 검증 통과!")

print("\n=== 모든 검증 완료: 100% 성공! ===")

File: scratch/test_calc.dart
import '../lib/core/data/occupation_data.dart';
import '../lib/core/tax_engine/freelancer_tax.dart';

void main() {
  print('=== 프리랜서 계산 엔진 테스트 시작 ===');

  // 시나리오 1: 1인미디어창작자 (940909)
  // 누적수입: 3,000만원, 개월수: 6개월, 부양가족: 0명
  // 예상 연소득: 6,000만원
  // 연 추정 단순경비: 4,000만 * 64.1% + 2,000만 * 49.7% = 2,564만 + 994만 = 3,558만원
  // 추정 소득금액: 6,000만 - 3,558만 = 2,442만원
  // 소득공제(본인): 150만원
  // 과세표준: 2,292만원
  // 국세 산출세액: 2,292만 * 15% - 126만 = 217.8만원
  // 지방세 산출세액: 217.8만 * 10% = 21.78만원
  // 연간 추정 총결정세액: 217.8만 + 21.78만 = 2,395,800원 (각각 10원 단위 절사하여 합산)
  // 연간 추정 기납부 3.3% 세액: 6,000만 * 3.3% = 198만원
  // 차감 예상액: 198만 - 2,395,800원 = -415,800원 (추가 납부)
  // 남은 개월수: 6개월 -> 월 권장 저축액: 415,800원 / 6 = 69,300원
  
  final result1 = FreelancerTaxCalculator.calculateTaxSimulation(
    accumulatedIncome: 30000000,
    inputMonths: 6,
    allowanceCount: 0,
    occupationCode: '940909',
  );

  print('\n[시나리오 1 검증: 1인미디어창작자 (940909)]');
  print('업종명: ${result1.occupationName}');
  print('연환산 추정 수입: ${result1.annualEstimatedIncome.toInt()} 원 (기대값: 60000000)');
  print('추정 필요경비: ${result1.estimatedExpense.toInt()} 원 (기대값: 35580000)');
  print('추정 사업소득금액: ${result1.estimatedBusinessIncome.toInt()} 원 (기대값: 24420000)');
  print('추정 과세표준: ${result1.taxBase.toInt()} 원 (기대값: 22920000)');
  print('연 추정 종합소득세 (국세): ${result1.annualIncomeTax.toInt()} 원 (기대값: 2178000)');
  print('연 추정 지방소득세 (지방세): ${result1.annualLocalTax.toInt()} 원 (기대값: 217800)');
  print('연환산 기납부 3.3% 세액: ${result1.annualEstimatedTotalWithholding.toInt()} 원 (기대값: 1980000)');
  print('최종 예상 결과: ${result1.expectedRefundOrPayment.toInt()} 원 (기대값: -415800)');
  print('월 세금 비축 저축액: ${result1.monthlyReserve.toInt()} 원 (기대값: 69300)');
  print('넛지 메시지: ${result1.reserveNudgeMessage}');

  assert(result1.annualEstimatedIncome == 60000000.0);
  assert(result1.estimatedExpense == 35580000.0);
  assert(result1.estimatedBusinessIncome == 24420000.0);
  assert(result1.taxBase == 22920000.0);
  assert(result1.annualIncomeTax == 2178000.0);
  assert(result1.annualLocalTax == 217800.0);
  assert(result1.expectedRefundOrPayment == -415800.0);
  assert(result1.monthlyReserve == 69300.0);
  print('=> 시나리오 1 검증 통과!');


  // 시나리오 2: 기타자영업 (940918)
  // 누적수입: 1,000만원, 개월수: 3개월, 부양가족: 1명 (본인 포함 2명 공제: 300만원)
  // 예상 연소득: 4,000만원
  // 연 추정 단순경비: 4,000만 * 79.4% = 3,176만원
  // 추정 소득금액: 4,000만 - 3,176만 = 824만원
  // 소득공제(본인+1인): 300만원
  // 과세표준: 524만원
  // 국세 산출세액: 524만 * 6% = 314,400원
  // 지방세 산출세액: 31,440원 (314,400원 * 10% 원화 절사)
  // 연간 추정 총결정세액: 314,400 + 31,440 = 345,840원
  // 연간 추정 기납부 3.3% 세액: 4,000만 * 3.3% = 132만원 (국세 120만, 지방세 12만)
  // 차감 예상액: 132만 - 345,840원 = 974,160원 (환급)
  
  final result2 = FreelancerTaxCalculator.calculateTaxSimulation(
    accumulatedIncome: 10000000,
    inputMonths: 3,
    allowanceCount: 1,
    occupationCode: '940918',
  );

  print('\n[시나리오 2 검증: 기타자영업 (940918)]');
  print('업종명: ${result2.occupationName}');
  print('연환산 추정 수입: ${result2.annualEstimatedIncome.toInt()} 원 (기대값: 40000000)');
  print('추정 필요경비: ${result2.estimatedExpense.toInt()} 원 (기대값: 31760000)');
  print('추정 사업소득금액: ${result2.estimatedBusinessIncome.toInt()} 원 (기대값: 8240000)');
  print('추정 과세표준: ${result2.taxBase.toInt()} 원 (기대값: 5240000)');
  print('연 추정 종합소득세 (국세): ${result2.annualIncomeTax.toInt()} 원 (기대값: 314400)');
  print('연 추정 지방소득세 (지방세): ${result2.annualLocalTax.toInt()} 원 (기대값: 31440)');
  print('최종 예상 결과: ${result2.expectedRefundOrPayment.toInt()} 원 (기대값: 974160)');
  print('넛지 메시지: ${result2.reserveNudgeMessage}');

  assert(result2.annualEstimatedIncome == 40000000.0);
  assert(result2.estimatedExpense == 31760000.0);
  assert(result2.estimatedBusinessIncome == 8240000.0);
  assert(result2.taxBase == 5240000.0);
  assert(result2.annualIncomeTax == 314400.0);
  assert(result2.annualLocalTax == 31440.0);
  assert(result2.expectedRefundOrPayment == 974160.0);
  print('=> 시나리오 2 검증 통과!');

  print('\n=== 모든 검증 완료: 100% 성공! ===');
}
